File: code-push-jd/ekyc/android-jetpack-compose/EKYCViewModel.kt
package com.example.ekyc.viewmodel

import android.app.Application
import android.nfc.Tag
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.ekyc.nfc.ChipData
import com.example.ekyc.nfc.NFCReader
import com.example.ekyc.security.SecureStorage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.jmrtd.BACKey

/**
 * ViewModel for eKYC flow
 * Orchestrates NFC reading, OCR scanning, and secure storage
 */
class EKYCViewModel(application: Application) : AndroidViewModel(application) {

    private val secureStorage = SecureStorage(application)
    
    private val _uiState = MutableStateFlow<EKYCUiState>(EKYCUiState.Initial)
    val uiState: StateFlow<EKYCUiState> = _uiState.asStateFlow()

    private val _mrzData = MutableStateFlow<MRZData?>(null)
    val mrzData: StateFlow<MRZData?> = _mrzData.asStateFlow()

    private val _chipData = MutableStateFlow<ChipData?>(null)
    val chipData: StateFlow<ChipData?> = _chipData.asStateFlow()

    /**
     * Process MRZ data from OCR scan
     */
    fun onMRZScanned(mrzData: MRZData) {
        _mrzData.value = mrzData
        _uiState.value = EKYCUiState.MRZScanned(mrzData)
    }

    /**
     * Start NFC reading with scanned MRZ data
     */
    fun readNFCChip(nfcReader: NFCReader, tag: Tag) {
        val mrz = _mrzData.value
        if (mrz == null) {
            _uiState.value = EKYCUiState.Error("Please scan MRZ first")
            return
        }

        viewModelScope.launch {
            _uiState.value = EKYCUiState.ReadingNFC

            // Create BAC key from MRZ
            val bacKey = BACKey(
                mrz.documentNumber,
                mrz.dateOfBirth,
                mrz.dateOfExpiry
            )

            val result = nfcReader.readChipData(tag, bacKey)
            
            result.onSuccess { data ->
                _chipData.value = data
                
                // Verify MRZ matches chip data
                if (verifyDataConsistency(mrz, data)) {
                    // Securely store sensitive data
                    secureStorage.storeChipData(data)
                    _uiState.value = EKYCUiState.Success(data)
                } else {
                    _uiState.value = EKYCUiState.Error("Data mismatch between MRZ and chip")
                }
            }.onFailure { error ->
                _uiState.value = EKYCUiState.Error(error.message ?: "NFC read failed")
            }
        }
    }

    /**
     * Verify consistency between scanned MRZ and chip data
     */
    private fun verifyDataConsistency(mrz: MRZData, chip: ChipData): Boolean {
        return mrz.documentNumber == chip.documentNumber &&
               mrz.dateOfBirth == chip.dateOfBirth &&
               mrz.dateOfExpiry == chip.dateOfExpiry
    }

    /**
     * Submit eKYC data to backend
     */
    fun submitEKYC() {
        val chip = _chipData.value
        if (chip == null) {
            _uiState.value = EKYCUiState.Error("No chip data available")
            return
        }

        viewModelScope.launch {
            _uiState.value = EKYCUiState.Submitting

            try {
                // TODO: Implement actual API call with mutual TLS
                // val response = ekycRepository.submit(chip)
                
                // Clear sensitive data after submission
                clearSensitiveData()
                
                _uiState.value = EKYCUiState.Submitted
            } catch (e: Exception) {
                _uiState.value = EKYCUiState.Error(e.message ?: "Submission failed")
            }
        }
    }

    /**
     * Clear all sensitive data from memory and storage
     */
    private fun clearSensitiveData() {
        _mrzData.value = null
        _chipData.value = null
        secureStorage.clearAll()
    }

    /**
     * Reset to initial state
     */
    fun reset() {
        clearSensitiveData()
        _uiState.value = EKYCUiState.Initial
    }

    override fun onCleared() {
        super.onCleared()
        clearSensitiveData()
    }
}

/**
 * UI States for eKYC flow
 */
sealed class EKYCUiState {
    object Initial : EKYCUiState()
    object ScanningMRZ : EKYCUiState()
    data class MRZScanned(val data: MRZData) : EKYCUiState()
    object ReadingNFC : EKYCUiState()
    data class Success(val chipData: ChipData) : EKYCUiState()
    object Submitting : EKYCUiState()
    object Submitted : EKYCUiState()
    data class Error(val message: String) : EKYCUiState()
}

/**
 * MRZ data from OCR scan
 */
data class MRZData(
    val documentNumber: String,
    val dateOfBirth: String,
    val dateOfExpiry: String,
    val firstName: String,
    val lastName: String
)

File: code-push-jd/ekyc/flutter/ekyc_viewmodel.dart
import 'package:flutter/foundation.dart';
import 'nfc_reader.dart';
import 'secure_storage.dart';

/// ViewModel for eKYC flow using ChangeNotifier
/// Orchestrates NFC reading, data verification, and submission
class EKYCViewModel extends ChangeNotifier {
  final NFCReader _nfcReader = NFCReader();
  final SecureStorage _secureStorage = SecureStorage();

  EKYCState _state = EKYCState.initial();
  MRZData? _mrzData;
  ChipData? _chipData;

  EKYCState get state => _state;
  MRZData? get mrzData => _mrzData;
  ChipData? get chipData => _chipData;

  /// Process scanned MRZ data
  void onMRZScanned(MRZData data) {
    _mrzData = data;
    _state = EKYCState.mrzScanned(data);
    notifyListeners();
  }

  /// Start NFC reading with scanned MRZ
  Future<void> readNFCChip() async {
    if (_mrzData == null) {
      _state = EKYCState.error('Please scan MRZ first');
      notifyListeners();
      return;
    }

    try {
      _state = EKYCState.readingNFC();
      notifyListeners();

      final mrzKey = MRZKey(
        documentNumber: _mrzData!.documentNumber,
        dateOfBirth: _mrzData!.dateOfBirth,
        dateOfExpiry: _mrzData!.dateOfExpiry,
      );

      final chipData = await _nfcReader.startReading(mrzKey);

      // Verify data consistency
      if (_verifyDataConsistency(_mrzData!, chipData)) {
        _chipData = chipData;

        // Store securely
        await _secureStorage.storeChipData(chipData);

        _state = EKYCState.success(chipData);
      } else {
        _state = EKYCState.error('Data mismatch between MRZ and chip');
      }
    } catch (e) {
      _state = EKYCState.error(e.toString());
    } finally {
      notifyListeners();
    }
  }

  /// Submit eKYC data to backend
  Future<void> submitEKYC() async {
    if (_chipData == null) {
      _state = EKYCState.error('No chip data available');
      notifyListeners();
      return;
    }

    try {
      _state = EKYCState.submitting();
      notifyListeners();

      // TODO: Implement actual API call with mutual TLS
      await Future.delayed(const Duration(seconds: 2)); // Simulate API call

      // Clear sensitive data after submission
      await _clearSensitiveData();

      _state = EKYCState.submitted();
    } catch (e) {
      _state = EKYCState.error('Submission failed: $e');
    } finally {
      notifyListeners();
    }
  }

  /// Reset to initial state
  Future<void> reset() async {
    await _clearSensitiveData();
    _state = EKYCState.initial();
    notifyListeners();
  }

  /// Check if NFC is available
  Future<bool> isNFCAvailable() => _nfcReader.isNFCAvailable();

  /// Check if NFC is enabled
  Future<bool> isNFCEnabled() => _nfcReader.isNFCEnabled();

  /// Verify consistency between MRZ and chip data
  bool _verifyDataConsistency(MRZData mrz, ChipData chip) {
    return mrz.documentNumber == chip.documentNumber &&
        mrz.dateOfBirth == chip.dateOfBirth &&
        mrz.dateOfExpiry == chip.dateOfExpiry;
  }

  /// Clear sensitive data from memory and storage
  Future<void> _clearSensitiveData() async {
    _mrzData = null;
    _chipData = null;
    await _secureStorage.clearAll();
  }

  @override
  void dispose() {
    _clearSensitiveData();
    super.dispose();
  }
}

/// UI States for eKYC flow
class EKYCState {
  final EKYCStateType type;
  final String? errorMessage;
  final MRZData? mrzData;
  final ChipData? chipData;

  const EKYCState._({
    required this.type,
    this.errorMessage,
    this.mrzData,
    this.chipData,
  });

  factory EKYCState.initial() =>
      const EKYCState._(type: EKYCStateType.initial);

  factory EKYCState.scanningMRZ() =>
      const EKYCState._(type: EKYCStateType.scanningMRZ);

  factory EKYCState.mrzScanned(MRZData data) =>
      EKYCState._(type: EKYCStateType.mrzScanned, mrzData: data);

  factory EKYCState.readingNFC() =>
      const EKYCState._(type: EKYCStateType.readingNFC);

  factory EKYCState.success(ChipData data) =>
      EKYCState._(type: EKYCStateType.success, chipData: data);

  factory EKYCState.submitting() =>
      const EKYCState._(type: EKYCStateType.submitting);

  factory EKYCState.submitted() =>
      const EKYCState._(type: EKYCStateType.submitted);

  factory EKYCState.error(String message) =>
      EKYCState._(type: EKYCStateType.error, errorMessage: message);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is EKYCState &&
          runtimeType == other.runtimeType &&
          type == other.type;

  @override
  int get hashCode => type.hashCode;
}

enum EKYCStateType {
  initial,
  scanningMRZ,
  mrzScanned,
  readingNFC,
  success,
  submitting,
  submitted,
  error,
}

/// MRZ data from OCR scan
class MRZData {
  final String documentNumber;
  final String dateOfBirth;
  final String dateOfExpiry;
  final String firstName;
  final String lastName;

  const MRZData({
    required this.documentNumber,
    required this.dateOfBirth,
    required this.dateOfExpiry,
    required this.firstName,
    required this.lastName,
  });

  Map<String, String> toMap() => {
        'documentNumber': documentNumber,
        'dateOfBirth': dateOfBirth,
        'dateOfExpiry': dateOfExpiry,
        'firstName': firstName,
        'lastName': lastName,
      };

  @override
  String toString() {
    return 'MRZData(documentNumber: $documentNumber, '
        'name: $lastName $firstName)';
  }
}
